### PR TITLE
[API] Add partition-by option to list APIs for feature-sets and feature-vectors

### PR DESCRIPTION
**Pull request by theSaarco**
Adding to the list APIs the option to partition the results according to the `name` field. The user can specify the ordering type (`asc` or `desc`) and select how many rows to return per each partition.
The immediate use-case for this is for the UI to retrieve a list of all feature-sets, and for each feature-set get only the latest instance of that object. This will be done by using `partition-by` and asking for a single row per partition.
The new parameters are:
* `partition-by` — a name of a field to group by (currently must be `name`). No default. If this parameter is specified, the grouping functionality is turned on.
* `rows-per-partition` — a number indicating the maximum number of items in a group. Default: 1.
* `partition-sort-by` — a name of a field to sort by. No default. Currently must be `updated`.
* `partition-order` — Default: `desc`

Functionality added to the REST interface, as well as to `RunDB` interface.
Future PRs will add support for these API options for Features and Entities list APIs as well.

### CODE REVIEW

File: mlrun/api/db/sqldb/db.py [hunk 1]
```diff
@@ -1181,6 +1181,36 @@ def list_entities(
                 )
         return schemas.EntitiesOutput(entities=entities_results)
 
+    @staticmethod
+    def _assert_group_by_parameters(group_by, sort):
+        if sort is None:
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                "sort parameter must be provided when group_by is used."
+            )
```
Comment (Hedingber) [inline mlrun/api/db/sqldb/db.py:1189]: I think we can have a sane default to `updated` (when group_by is given)

File: mlrun/api/db/sqldb/db.py [hunk 2]
```diff
@@ -1181,6 +1181,36 @@ def list_entities(
...
+            .label("row_number")
+        )
+
+        # Need to generate a subquery so we can filter based on the row_number, since it
+        # is a window function using over().
+        subquery = query.add_column(row_number_column).subquery()
+        return session.query(aliased(cls, subquery)).filter(
+            subquery.c.row_number <= rows_per_group
+        )
```
Comment (Hedingber) [inline mlrun/api/db/sqldb/db.py:1212]: I hope you tested it well, cause that is Chinese to me 😅

File: mlrun/api/schemas/constants.py [hunk 3]
```diff
@@ -52,3 +52,30 @@ class HeaderNames:
...
+
+class GroupByField(str, Enum):
+    name = "name"  # Supported for feature-store objects
+    key = "key"  # Supported for artifacts
+
+    def to_group_by_db_field(self, db_cls):
+        if self.value == GroupByField.name:
+            return db_cls.name
+        else:
+            return db_cls.key
```
Comment (Hedingber) [inline mlrun/api/schemas/constants.py:65]: I think we better have a per object enum class

File: tests/api/api/feature_store/base.py [hunk 4]
```diff
@@ -60,3 +60,63 @@ def _assert_diff_as_expected_except_for_specific_metadata(
...
+        object_name,
+        project_name,
+        "group-by=name&sort-by=updated&rows-per-group=1&order=desc",
+        count,
+    )[object_name]
+
+    for result_object in results:
+        assert result_object["metadata"]["tag"] == "newest"
+
+    results = _list_and_assert_objects(
```
Comment (Hedingber) [inline tests/api/api/feature_store/base.py:80]: Add some comments before each case cause it's hard to see by eye what has changed, for example here I would add `"# testing desc`

**Review by Hedingber — CHANGES_REQUESTED** [review]

File: mlrun/api/db/sqldb/db.py (hunk 2 above)
Comment (theSaarco) [reply mlrun/api/db/sqldb/db.py:1212]: It's not THAT complicated, will be happy to walk you through the details. And, yes.

File: mlrun/api/db/sqldb/db.py (hunk 1 above)
Comment (theSaarco) [reply mlrun/api/db/sqldb/db.py:1189]: We agreed to not provide a default for the sort field (we may want to add `created` shortly). I would prefer keeping it without a default and having the user specify the value.

File: mlrun/api/schemas/constants.py (hunk 3 above)
Comment (theSaarco) [reply mlrun/api/schemas/constants.py:65]: Well, currently `artifacts` code is not in place yet, so I see no point in defining the enum for `artifacts`. I will change this to be `FeatureStoreGroupByField` and remove the `key` option. I don't see much point in having a different enum for `FeatureSet` and for `FeatureStore` as a lot of the code treats them the same (which is part of the point of the way they're built anyway).

**Review by Hedingber — APPROVED** [review]

**Review by Hedingber — APPROVED** [review]

**Pull request merged**